src/kvstore/gpu_topology.h

```diff
@@ -0,0 +1,1061 @@
+      }
+    }
+  }
+
+  // Check that all GPUs have at least 1 NVLink connection
+  int max_value = 0;
+  for (unsigned int i = 0; i < max.size(); ++i) {
+    if (max[i] > max_value)
+      max_value = max[i];
+  }
+
+  // If all GPUs are connected by NVLink, then we can use NVLink only
+  // to communicate instead of going over PCI-E, so we set PCI-E links to 0
+  //
+  // Otherwise, we will make distinction between PCI-E GPUDirect links and
+  // PCI-E through CPU links, which are slower and show queueing effect (i.e.
+  // The most packets there are, the slower).
+  //
+  // For the latter links, we will set links that were 0 to 1/num_gpus to
+  // account for this queuing effect.
+  bool connected = IsConnected(*matrix, num_gpus);
+
+  if (connected) {
+    for (auto& matrix_value : *matrix) {
+      matrix_value = (matrix_value == 1) ? 0 : matrix_value;
+    }
+  } else {
+    for (auto& matrix_value : *matrix) {
+      matrix_value = (matrix_value == 1) ? 1./num_gpus : matrix_value;
+    }
+  }
+
+  if (kLogTree)
+    PrintMatrix("Weight W", *matrix, num_gpus, num_gpus);
+#else
+  LOG(WARNING) << "GPU required for link topology";
+#endif
+}
+
+// Dense matrix-vector multiplication
+// Assume: matrix is square
+//   y = A*x (no accumulate)
+template <typename T>
+inline void gemv(const std::vector<T>& A, const std::vector<int>& x,
+                 std::vector<T>* y) {
+  int nrows = x.size();
+  int count = 0;
+  for (int row=0; row < nrows; ++row) {
+    (*y)[row] = 0;
+    for (int col=0; col < nrows; ++col) {
+      (*y)[row] += A[count]*static_cast<T>(x[col]);
+      count++;
+    }
+  }
+}
+
+// Element-wise multiplication between 2 dense vectors
+//   w = w * alpha*u
+template <typename T>
+inline void ewisemult(const std::vector<int>& u, T alpha, std::vector<T>* w) {
+  int nelem = u.size();
+  for (int i=0; i < nelem; ++i) {
+    (*w)[i] *= alpha*static_cast<T>(u[i]);
+  }
+}
+
+// Computes best 2 nodes a,b to swap given objective function:
+//   g = max_{a \in A, b \in B} D(a) + D(b) - 2*W(a,b)
+//
+// Optimization: Only need to look at upper triangular since weight matrix is
+//   symmetric
+template <typename T>
+inline void FindBestMove(const std::vector<T>& W,
+                         const std::vector<int>& P_temp,
+                         const std::vector<T>& D,
+                         const std::unordered_set<int>& used,
+                         int* a, int* b, T* g) {
+  int nrows = P_temp.size();
+  *g = 0;
+  *a = -1;
+  *b = -1;
+  for (int row=0; row < nrows; ++row) {
+    if (P_temp[row] == 0 || used.find(row) != used.end()) continue;
+    for (int col=row+1; col < nrows; ++col) {
+      if (P_temp[col] == 0 || P_temp[row] == P_temp[col]) continue;
+
+      T cost = D[row]+D[col]-2*W[row*nrows+col];
+      if (cost > *g) {
+        *g = cost;
+        *a = row;
+        *b = col;
+      }
+    }
+  }
+}
+
+// Performs partition on each existing partition in graph W if partition has
```

Comment:
Thanks for the great comments for the functions, but could you fix comment style to how it's standard in the codebase. Here's an example https://github.com/apache/incubator-mxnet/blob/b4156da26cfe741619227ae726872b1255194900/src/kvstore/kvstore_utils.h#L37